### ISSUE TITLE
[BUG] Potentially unsafe releasing of the GIL in `DeviceBuffer`

### ISSUE THREAD

**Issue by shwina** [rapidsai/rmm#944]
## The problem

When allocating memory in the constructor of `DeviceBuffer`, we release the GIL. See [here](https://github.com/rapidsai/rmm/blob/5584a0ccdb4d8d68df96eb6e8121a3baa4f4e168/python/rmm/_lib/device_buffer.pyx#L83-L98).

After the allocation, [we store a reference to the current memory resource](https://github.com/rapidsai/rmm/blob/5584a0ccdb4d8d68df96eb6e8121a3baa4f4e168/python/rmm/_lib/device_buffer.pyx#L97) associated with the allocation, so that the memory resource does not get deleted before the DeviceBuffer.

However, as @viclafargue pointed out in an offline discussion, when the GIL is released, another Python thread could change the current memory resource, via a call to [`set_current_device_resource`](https://github.com/rapidsai/rmm/blob/5584a0ccdb4d8d68df96eb6e8121a3baa4f4e168/python/rmm/_lib/memory_resource.pyx#L779) for example. We can thus end up with a situation where we hold a reference to a different memory resource than was used for the allocation. The memory resource that was actually used for the allocation could then be deleted before the DeviceBuffer gets destructed, leading to a segfault during destruction.

## Solutions

There are two potential solutions I can think of:

1. We don't release the GIL in the constructor of `DeviceBuffer`.
2. We still release the GIL, but we store a reference to the current memory resource _first_. Then, we explicitly pass that memory resource to the constructor of the underlying `rmm::device_buffer`. Something along these lines:

     ```python
        # Save a reference to the MR and stream used for allocation
        self.mr = get_current_device_resource()
        self.stream = stream

        with nogil:
            c_ptr = <const void*>ptr

            if size == 0:
                self.c_obj.reset(new device_buffer(self.mr.c_obj.get()))
            elif c_ptr == NULL:
                self.c_obj.reset(new device_buffer(size, stream.view(), self.mr.c_obj.get()))
            else:
                self.c_obj.reset(new device_buffer(c_ptr, size, stream.view(), self.mr.c_obj.get()))

                if stream.c_is_default():
                    stream.c_synchronize()
    ```

(2) seems preferred, unless I'm overlooking something subtle. Will open a PR to address.

**Comment by jakirkham** [comment]
+1 to option 2

**Comment by harrism** [comment]
On the C++ side, a device_buffer takes an explicit memory resource in any of its methods that allocates memory  (e.g. the constructor), and it stores it so that it uses the same MR for deallocation. In C++ we don't call `get_current_device_resource` in `device_buffer`, for exactly this reason.  

Is there a way you could make it explicit in Python as well, rather than using get_current_device_resource? If the user wants to use `get_current_device_resource`, they could call it themselves and pass the returned resource to the device buffer.

**Comment by jakirkham** [comment]
I think it makes sense to have the option of specifying a resource to use when allocating a `DeviceBuffer` (we don't have that currently). Possibly worth doing at the same time.

That said, there is still value in having a default case that falls back to `get_current_device_resource`. Given `DeviceBuffer` is used all over the place, changing how it gets called would be too disruptive.

**Comment by vyasr** [comment]
I agree, making the memory resource an optional constructor parameter would be useful. I don't think we should get rid of the default behavior though, at least not in the short term.

**Comment by harrism** [comment]
Can we make it look like the C++ API (the default value of the optional parameter is a call to `get_current_device_resource)? That way it's very clear from the interface what the default behavior is.

https://github.com/rapidsai/rmm/blob/5584a0ccdb4d8d68df96eb6e8121a3baa4f4e168/include/rmm/device_buffer.hpp#L130-L133

**Comment by jakirkham** [comment]
We should be able to put it in the same position as we have tried to follow that signature thus far.

While I understand the suggestion to include `get_current_device_resource` in the function signature and agree it makes sense conceptually, in Python [unfortunately this can result in bad behavior as it is only evaluated once]( https://towardsdatascience.com/python-pitfall-mutable-default-arguments-9385e8265422 ). So the typical way to handle this in Python would be to assign `None` to this argument, then check if it is `None` and replace that inside the function.

That said, we can clarify this behavior further in the docstring.

**Comment by harrism** [comment]
@shwina should we bump to next release?

**Comment by shwina** [comment]
Just did so - thanks!

**Comment by wence-** [comment]
This was closed by #1514
